pkg/cloud/services/compute/instance.go

```diff
@@ -407,68 +386,129 @@ func (s *Service) getVolumeByName(name string) (*volumes.Volume, error) {
 	if err != nil {
+		record.Eventf(eventObject, "FailedCreateVolume", "Failed to create volume; name=%s size=%d err=%v", opts.Name, opts.Size, err)
 		return nil, err
 	}
-	if volume != nil {
-		if volume.Size != size {
-			return nil, fmt.Errorf("exected to find volume %s with size %d; found size %d", name, size, volume.Size)
+	record.Eventf(eventObject, "SuccessfulCreateVolume", "Created volume; id=%s", createdVolume.ID)
+	return createdVolume, err
+}
+
+func (s *Service) waitForVolume(volumeID string, timeout time.Duration, retryInterval time.Duration) error {
+	return wait.PollUntilContextTimeout(context.TODO(), retryInterval, timeout, true, func(_ context.Context) (bool, error) {
+		volume, err := s.getVolumeClient().GetVolume(volumeID)
+		if err != nil {
+			if capoerrors.IsRetryable(err) {
+				return false, nil
+			}
+			return false, err
 		}
 
-		s.scope.Logger().V(3).Info("Using existing root volume", "name", name)
-		return volume, nil
-	}
+		switch volume.Status {
+		case "available":
+			return true, nil
+		case "error":
+			return false, fmt.Errorf("volume %s is in error state", volumeID)
+		default:
+			return false, nil
+		}
+	})
+}
 
+func (s *Service) getOrCreateVolumeBuilder(eventObject runtime.Object, instanceSpec *InstanceSpec, blockDevice infrav1.AdditionalBlockDevice, imageID string, description string) (*volumes.Volume, error) {
 	availabilityZone := instanceSpec.FailureDomain
-	if rootVolume.AvailabilityZone != "" {
-		availabilityZone = rootVolume.AvailabilityZone
+	if blockDevice.AvailabilityZone != "" {
+		availabilityZone = blockDevice.AvailabilityZone
 	}
 
 	createOpts := volumes.CreateOpts{
-		Size:             rootVolume.Size,
-		Description:      fmt.Sprintf("Root volume for %s", instanceSpec.Name),
-		Name:             rootVolumeName(instanceSpec.Name),
+		Name:             volumeName(instanceSpec.Name, blockDevice.NameSuffix),
+		Description:      description,
+		Size:             blockDevice.Size,
 		ImageID:          imageID,
 		Multiattach:      false,
 		AvailabilityZone: availabilityZone,
-		VolumeType:       rootVolume.VolumeType,
+		VolumeType:       blockDevice.VolumeType,
 	}
-	volume, err = s.getVolumeClient().CreateVolume(createOpts)
-	if err != nil {
-		record.Eventf(eventObject, "FailedCreateVolume", "Failed to create root volume; size=%d imageID=%s err=%v", size, imageID, err)
-		return nil, err
-	}
-	record.Eventf(eventObject, "SuccessfulCreateVolume", "Created root volume; id=%s", volume.ID)
-	return volume, err
+
+	return s.getOrCreateVolume(eventObject, createOpts)
 }
 
-// applyRootVolume sets a root volume if the root volume Size is not 0.
-func applyRootVolume(opts servers.CreateOptsBuilder, volume *volumes.Volume) servers.CreateOptsBuilder {
-	if volume == nil {
-		return opts
+func (s *Service) getBlockDevices(eventObject runtime.Object, instanceSpec *InstanceSpec, imageID string, timeout time.Duration, retryInterval time.Duration) ([]bootfromvolume.BlockDevice, error) {
```

Comment:
Please add a doc string for this function

Reply:
@EmilienM what about this one?

Reply:
done